src/faivor/metrics_api.py

```diff
@@ -0,0 +1,600 @@
+import json
+import numpy as np
+import pandas as pd
+import logging
+from pathlib import Path
+from typing import Dict, List, Any, Union, Optional, Tuple
+from sklearn.metrics import roc_curve, roc_auc_score, precision_recall_curve, average_precision_score, confusion_matrix
+
+from faivor.model_metadata import ModelMetadata
+from faivor.parse_data import detect_delimiter
+from faivor.metrics.regression import metrics as regression_metrics
+from faivor.metrics.classification import metrics as classification_metrics
+from faivor.utils import convert_to_json_serializable, safe_divide
+
+logger = logging.getLogger(__name__)
+
+class MetricsCalculator:
+    """
+    Calculate metrics for model predictions and perform subgroup analysis.
+    """
+    
+    def __init__(self, model_metadata: ModelMetadata, predictions: List, 
+                expected_outputs: List, inputs: Optional[List] = None):
+        """
+        Initialize the metrics calculator.
+        
+        Parameters
+        ----------
+        model_metadata : ModelMetadata
+            Model metadata containing information about the model.
+        predictions : List
+            Predictions from the model.
+        expected_outputs : List
+            Expected outputs (ground truth) for evaluation.
+        inputs : List, optional
+            Input data used for making predictions, by default None.
+        """
+        self.model_metadata = model_metadata
+        self.predictions = predictions
+        self.expected_outputs = expected_outputs
+        self.inputs = inputs
+        self.output_field = model_metadata.output
+        self.metadata_json = model_metadata.metadata
+        
+    def prepare_data(self) -> Tuple[np.ndarray, np.ndarray, List[int]]:
+        """
+        Prepare data for metrics calculation by filtering valid samples.
+        
+        Returns
+        -------
+        tuple
+            Tuple containing (y_true, y_pred, valid_indices)
+        """
+        valid_indices = []
+        y_true_values = []
+        y_pred_values = []
+        
+        for i, (output, pred) in enumerate(zip(self.expected_outputs, self.predictions)):
+            try:
+                # need to check if expected output has the required key and is numeric
+                if isinstance(output, dict) and self.output_field in output:
+                    true_val = output[self.output_field]
+                    if true_val in ('x', 'X', ''):
+                        continue
+                    
+                    true_float = float(true_val)
+                    pred_float = float(pred)
+                    
+                    valid_indices.append(i)
+                    y_true_values.append(true_float)
+                    y_pred_values.append(pred_float)
+                else:
+                    continue
+            except (ValueError, TypeError):
```

Comment:
```suggestion
            except (ValueError, TypeError):
```
The `else` clause can be omited here